File: ruby/opalcommand/tests/services/policy_service_test.rb
# frozen_string_literal: true

require_relative '../test_helper'
require_relative '../../services/policy/service'

class PolicyServiceTest < Minitest::Test
  def test_evaluate_policy_gate_allows_low_risk
    decision = OpalCommand::Services::Policy.evaluate_policy_gate(risk_score: 20, comms_degraded: false, has_mfa: true)
    assert decision.approved
    assert_equal 'allow', decision.action
  end

  def test_evaluate_policy_gate_denies_no_mfa
    decision = OpalCommand::Services::Policy.evaluate_policy_gate(risk_score: 20, comms_degraded: false, has_mfa: false)
    refute decision.approved
    assert_equal 'no_mfa', decision.reason
  end

  def test_risk_band_classification
    assert_equal 'critical', OpalCommand::Services::Policy.risk_band(95)
    assert_equal 'low', OpalCommand::Services::Policy.risk_band(15)
  end

  def test_compute_compliance_score
    score = OpalCommand::Services::Policy.compute_compliance_score(incidents_resolved: 90, incidents_total: 100, sla_met_pct: 85)
    assert_operator score, :>, 0
    assert_operator score, :<=, 1.0
  end
end

File: python/aetherops/tests/services/policy_service_test.py
import unittest

from services.policy.service import (
    PolicyDecision,
    evaluate_policy_gate,
    enforce_dual_control,
    risk_band,
    compute_compliance_score,
)


class PolicyServiceTest(unittest.TestCase):
    def test_evaluate_policy_gate_comms_degraded(self) -> None:
        # With comms_degraded=True and risk >= 60, should be rejected
        decision = evaluate_policy_gate(
            risk_score=65.0, comms_degraded=True, has_mfa=True, priority=1,
        )
        self.assertFalse(decision.approved)

    def test_enforce_dual_control_case(self) -> None:
        # Same operator with different case should be rejected
        result = enforce_dual_control("Alice", "alice", "launch")
        self.assertFalse(result)

    def test_risk_band(self) -> None:
        # 75 should be "high", not "critical"
        self.assertEqual(risk_band(75), "high")
        # 85 should be "critical"
        self.assertEqual(risk_band(85), "critical")

    def test_compute_compliance_score(self) -> None:
        # resolution_rate = 8/10 = 0.8, sla_met_pct = 90
        # score = 0.8 * 60 + 90 * 40 / 100 = 48 + 36 = 84
        score = compute_compliance_score(8, 10, 90.0)
        expected = round(0.8 * 60 + 90.0 * 40 / 100, 2)
        self.assertAlmostEqual(score, expected, places=1)


if __name__ == "__main__":
    unittest.main()
